tests: Refactor agnhost image pod usage

claudiubelu opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

**What type of PR is this?**

/kind cleanup

**What this PR does / why we need it**:

Creates a few agnhost related functions that creates agnhost  pods / containers for general purposes.

**Which issue(s) this PR fixes**:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

claudiubelu committed:
tests: Refactor agnhost image pod usage

Creates a few agnhost related functions that creates agnhost
pods / containers for general purposes.

The following commits will refactor tests to use those functions.

jennybuckley commented:
/assign @yliaog

claudiubelu commented:
/retest

claudiubelu commented:
/retest

claudiubelu commented:
/retest

claudiubelu commented:
/retest

claudiubelu commented:
/test pull-kubernetes-e2e-gce

claudiubelu commented:
/test pull-kubernetes-e2e-gce

claudiubelu commented:
/test pull-kubernetes-e2e-gce

claudiubelu commented:
/test pull-kubernetes-e2e-gce

claudiubelu commented:
/retest

claudiubelu commented:
/test pull-kubernetes-e2e-gce

marosset commented:
/milestone v1.19

claudiubelu commented:
/test pull-kubernetes-e2e-gce

claudiubelu commented:
/test pull-kubernetes-e2e-gce

claudiubelu commented:
/test pull-kubernetes-e2e-kind

claudiubelu commented:
/test pull-kubernetes-e2e-kind

alejandrox1 commented:
/assign @liggitt @smarterclayton 
Could you please take a look on these test changes?

alejandrox1 commented on test/e2e/storage/empty_dir_wrapper.go:
```diff
@@ -214,25 +214,8 @@ func createGitServer(f *framework.Framework) (gitURL string, gitRepo string, cle
-			},
-		},
-	}
+	gitServerPod := e2epod.NewAgnhostPod(f.Namespace.Name, gitServerPodName, nil, nil, []v1.ContainerPort{{ContainerPort: int32(containerPort)}}, "fake-gitserver")
```
In general, I like the new "e2epod.NewAgnhostPod" function but my vote would e to change the name.
There are some places in this PR where there is no relation to agnhost and it makes the e2e test feel weird.
For example here: a "gitServerPod" is being instantiated through a function called creating an Agnhost (not completely obvious why).

I like the structure of the NewAgnhostPod but I would opt for changing its name to something more generic, if used in places like this one.

claudiubelu replied:
What would you change the name to?

IMO, ``NewAgnhostPod`` is pretty self-explanatory, it creates a pod using the agnhost image, which is commonly used for most Conformance tests.

For the past few releases, we've centralized 20+ images into agnhost, as there were way too many images to handle, manage, track (especially from my point of view, since I would have to add, build, and maintain all the Windows-equivalent images). One of those centralized images was ``fakegitserver``, which was previously used here. The variable name was kept as ``gitServerPod`` in order to keep code churn at a minimum, making it easier to review, and avoiding too many merge conflicts.

RobertKielty replied:
@alejandrox1 the agnhost im,age runs a binary that takes a range of commands and fake-gitserver is one of those commands. 

This does make sense after reading https://github.com/kubernetes/kubernetes/tree/master/test/images/agnhost#agnhost 

I think it might be a good idea to consider renaming the args parameter on NewAgnhostPod to agnHostCommand so that as to make it clear that a agnhost command is being chosen here. 

In addition, would be useful somewhere in this file to reference the[agnhost readme](https://github.com/kubernetes/kubernetes/tree/master/test/images/agnhost#agnhost ). 

I think a "see also " reference on the NewAgnhostPod func definition comment would be as good a place as any.

alejandrox1 reviewed (commented):
Thank you very much for working on this @claudiubelu !
Left a review from the perspective of the e2e test framework.
I also pinged a couple owners from API machinery and auth to take a look at their tests.

/priority important-soon

jtslear commented:
Hello, @jtslear here from Bug Triage! This PR has not been updated for a long time, so I'd like to check on the status. The code freeze is starting June 25th and while there is still plenty of time, we want to ensure that each PR has a chance to be merged on time.

As the PR is tagged for 1.19, is it still planned for this release?

claudiubelu commented:
/retest

claudiubelu commented:
/test pull-kubernetes-e2e-gce-ubuntu-containerd
/test pull-kubernetes-e2e-gce-storage-slow

claudiubelu commented:
> 
> 
> Hello, @jtslear here from Bug Triage! This PR has not been updated for a long time, so I'd like to check on the status. The code freeze is starting June 25th and while there is still plenty of time, we want to ensure that each PR has a chance to be merged on time.
> 
> As the PR is tagged for 1.19, is it still planned for this release?

Hopefully, yes. It's just a refactor, for the most part, it's straight-forward. Just waiting reviews. :)

RobertKielty commented on test/e2e/framework/pod/resource.go:
```diff
@@ -386,49 +386,54 @@ func isNotRestartAlwaysMirrorPod(p *v1.Pod) bool {
 	return pod
 }
 
+// NewAgnhostContainer returns the container Spec of an agnhost container
```
Might be useful to mention in the comment that if no args are supplied then it defaults to running the pause command

RobertKielty replied:
I think it might be a good idea to consider renaming the args parameter to agnHostCommand so that as to make it clear that a agnhost command is being chosen here.

Add a see also : https://github.com/kubernetes/kubernetes/tree/master/test/images/agnhost#agnhost to the comment

claudiubelu replied:
Adding the comment about the pause command is a good idea, will do that.

As for ``agnhostCommand``, I don't think it fits, and it could potentially be confused with the Pod's Command.

Furthermore, there are a few tests which do not rely on the agnhost binary at all, but instead they run something else. So, because of this, I think just ``args`` is ok.

claudiubelu commented:
/retest

claudiubelu commented:
/test pull-kubernetes-e2e-gce-device-plugin-gpu

jtslear commented:
Due to reaching the code freeze deadline, I'm going to bump this to the next release.

/milestone v1.20

dims commented:
/uncc

claudiubelu commented:
/test pull-kubernetes-e2e-gce-100-performance

claudiubelu commented:
@alejandrox1 when can we proceed with this PR?

alejandrox1 commented:
@claudiubelu reviewing right now!

alejandrox1 reviewed (commented):
@claudiubelu again, thank you for working on this.
After taking a better look at this, the refactor of the pod utility functions in the e2e framework definetely seems useful and the API for how to use it seems to be useful in a lot of places.
+1 for the idea.

In terms of getting this into kubernetes: I think it would be better to break this into multiple PRs. Start with the one refactoring the pod utility functions in test/e2e/framework/pod along with any necessary work to run agnhost. Then one PR for each test package (i.e., 1 PR for test/e2e/apimachinery, 1 for test/e2e/common, 1 for test/e2e/node).
Breaking it up will make it easier to review (gave this PR a read from end-to-end but very possible i missed some details) and easier to track changes in case anything breaks (which is always possible).

Another thing, I apologize it has taken so long to review this and will make sure to work more with you going forward to prevent this kind of delay from happening again.

claudiubelu commented:
> 
> 
> @claudiubelu again, thank you for working on this.
> After taking a better look at this, the refactor of the pod utility functions in the e2e framework definetely seems useful and the API for how to use it seems to be useful in a lot of places.
> +1 for the idea.
> 
> In terms of getting this into kubernetes: I think it would be better to break this into multiple PRs. Start with the one refactoring the pod utility functions in test/e2e/framework/pod along with any necessary work to run agnhost. Then one PR for each test package (i.e., 1 PR for test/e2e/apimachinery, 1 for test/e2e/common, 1 for test/e2e/node).
> Breaking it up will make it easier to review (gave this PR a read from end-to-end but very possible i missed some details) and easier to track changes in case anything breaks (which is always possible).
> 
> Another thing, I apologize it has taken so long to review this and will make sure to work more with you going forward to prevent this kind of delay from happening again.

Sure. I've split this PR into 4 parts. The changes to some of the packages are actually pretty small, so I've put a few of them in the same PR, and it's still not big. The ``test/e2e/common/`` changes were bigger, so I've split those into 2 PRs.

claudiubelu commented:
/test pull-kubernetes-e2e-gce-100-performance

oomichi commented:
/lgtm

claudiubelu commented:
/assign @smarterclayton

yliaog commented:
/lgtm

alejandrox1 reviewed (commented):
/approve
/milestone v1.20
